server/store.go

```diff
@@ -0,0 +1,840 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/larkox/mattermost-plugin-badges/badgesmodel"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/plugin"
+)
+
+var errInvalidBadge = errors.New("invalid badge")
+var errBadgeNotFound = errors.New("badge not found")
+
+type Store interface {
+	// Interface
+	GetUserBadges(userID string) ([]badgesmodel.UserBadge, error)
+	GetAllBadges() ([]badgesmodel.AllBadgesBadge, error)
+	GetBadgeDetails(badgeID badgesmodel.BadgeID) (*badgesmodel.BadgeDetails, error)
+
+	// Autocomplete
+	GetGrantSuggestions(user model.User) ([]badgesmodel.Badge, error)
+	GetEditTypeSuggestions(user model.User) (badgesmodel.BadgeTypeList, error)
+	GetTypeSuggestions(user model.User) (badgesmodel.BadgeTypeList, error)
+	GetEditBadgeSuggestions(user model.User) ([]badgesmodel.Badge, error)
+
+	// API
+	AddBadge(badge badgesmodel.Badge) (*badgesmodel.Badge, error)
```

Comment:
A lot of this methods are receiving objects but returning pointeres, also we usually work with `*model.User` because is what is returned from mattermost API so the need of dereferencing the pointer (and copy the data) to pass it to the methods is not necesarilly wrong, but looks weird in the context of mattermost code.

Do you have a good reason for that?

Reply:
Not sure if originally I had a better reason apart of "trying to avoid pointers". The return values as pointers were probably for the versatility of returning the object or nil.

In order to simplify the code, I have set almost everything as a pointer, since it is not needed to pass them as values.